community/bisect.go

```diff
@@ -0,0 +1,216 @@
+		s := math.Inf(-1)
+		for i := 0; i < effort; i++ {
+			s = math.Max(s, score(ModularizeMultiplex(g, weights, []float64{resolution}, all, src)))
+		}
+		return s
+	}
+}
+
+// Profile returns an approximate profile of score values in the resolution domain [low,high)
+// at the given granularity. The score is calculated by bisecting calls to fn. If log is true,
+// log space bisection is used, otherwise bisection is linear. The function fn should be
+// monotonically decreasing with increasing input in at least 1/grain samples. Profile will
+// attempt to detect non-monotonicity during the bisection.
+//
+// Since exact modularity optimization is known to be NP-hard and Profile calls modularization
+// routines repeatedly, it is unlikely to return the exact resolution profile.
+func Profile(fn func(float64) float64, log bool, grain, low, high float64) (profile []Domain, err error) {
+	if low >= high {
+		return nil, errors.New("community: zero or negative width domain")
+	}
+
+	defer func() {
+		r := recover()
+		e, ok := r.(nonDecreasing)
+		if ok {
+			err = e
+			return
+		}
+		if r != nil {
+			panic(r)
+		}
+	}()
+	profile = bisect(fn, log, grain, low, fn(low), high, fn(high))
+
+	// We may have missed some non-monotonicity,
+	// so merge low score discordant domains into
+	// their lower resolution neighbours.
+	return fixUp(profile), nil
+}
+
+type nonDecreasing int
+
+func (n nonDecreasing) Error() string {
+	return fmt.Sprintf("community: profile does not reliably monotonically decrease: tried %d times", n)
+}
+
+func bisect(fn func(float64) float64, log bool, grain, low, scoreLow, high, scoreHigh float64) []Domain {
+	if low >= high {
+		panic("community: zero or negative width domain")
+	}
+	if math.IsNaN(scoreLow) || math.IsNaN(scoreHigh) {
+		return nil
+	}
+
+	// Heuristically determine a reasonable number
+	// of times to try to get a higher value.
+	maxIter := int(1 / grain)
+
+	for n := 0; scoreLow < scoreHigh; n++ {
+		if n > maxIter {
+			panic(nonDecreasing(n))
+		}
+		scoreLow = fn(low)
+	}
+
+	if scoreLow == scoreHigh || tooSmall(low, high, grain, log) {
+		return []Domain{{Low: low, High: high, Score: scoreLow}}
+	}
+
+	var mid float64
+	if log {
+		mid = math.Sqrt(low * high)
+	} else {
+		mid = (low + high) / 2
+	}
+
+	scoreMid := math.Inf(-1)
+	for n := 0; scoreMid < scoreHigh; n++ {
+		if n > maxIter {
+			panic(nonDecreasing(n))
+		}
+		scoreMid = fn(mid)
+	}
+
+	lower := bisect(fn, log, grain, low, scoreLow, mid, scoreMid)
+	higher := bisect(fn, log, grain, mid, scoreMid, high, scoreHigh)
+	for n := 0; lower[len(lower)-1].Score < higher[0].Score; n++ {
+		if n > maxIter {
+			panic(nonDecreasing(n))
+		}
+		lower[len(lower)-1].Score = fn(low)
+	}
+
+	if lower[len(lower)-1].Score == higher[0].Score {
+		higher[0].Low = lower[len(lower)-1].Low
+		lower = lower[:len(lower)-1]
+		if len(lower) == 0 {
+			return higher
+		}
+	}
+	return append(lower, higher...)
+}
+
+// fixUp non-monotonically decreasing domains.
```

Comment:
Strictly speaking it's not the domains that should be monotonically decreasing but it's unexported and understandable anyway, so ...